{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d31fbd54",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from weblogo import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2576331b",
   "metadata": {},
   "source": [
    "### Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "66ddede0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sea = \"sea\"  # change to specific directory of MEME/bin/sea\n",
    "df_name = \"Mix_clustered.csv\"  # input, be careful for multiindex\n",
    "motif_col = \"('INFO', 'motif_F10')\"\n",
    "cluster_col = \"Cluster\"\n",
    "meme_fn = \"mixed.meme\"\n",
    "fasta_fn = \"mixed.fa\"\n",
    "sea_out = \"sea_out\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d387b7b7",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "78bddc68",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(df_name, index_col=[0,1,2], header=[0])\n",
    "# df = pd.read_csv(df_name, index_col=[0,1,2], header=[0, 1]) # for dual header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8afe2cc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>('INFO', 'motif_F10')</th>\n",
       "      <th>Human</th>\n",
       "      <th>Mouse</th>\n",
       "      <th>X. laevis</th>\n",
       "      <th>X. tropicalis</th>\n",
       "      <th>Zebrafish</th>\n",
       "      <th>Fly</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10@100177160@-</th>\n",
       "      <th>ENSG00000107521</th>\n",
       "      <th>HPS1</th>\n",
       "      <td>CTAGACCACCCTCCATGTCAG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.848524</td>\n",
       "      <td>8.074862</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10@100177994@-</th>\n",
       "      <th>ENSG00000107521</th>\n",
       "      <th>HPS1</th>\n",
       "      <td>TCCAGATGATCGAGGTGCCCG</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.548068</td>\n",
       "      <td>1.524192</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10@100183394@-</th>\n",
       "      <th>ENSG00000107521</th>\n",
       "      <th>HPS1</th>\n",
       "      <td>GGACTTCTTGCTGGTGAAGAG</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.139991</td>\n",
       "      <td>2.309283</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10@100183566@-</th>\n",
       "      <th>ENSG00000107521</th>\n",
       "      <th>HPS1</th>\n",
       "      <td>CAGCCCCCAGCAGGGGAGGCC</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.001106</td>\n",
       "      <td>1.070967</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10@100184075@-</th>\n",
       "      <th>ENSG00000107521</th>\n",
       "      <th>HPS1</th>\n",
       "      <td>CCGGATCCTCCTGGGAGTGAG</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.508685</td>\n",
       "      <td>0.452521</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XHet@68785@-</th>\n",
       "      <th>FBgn0039945</th>\n",
       "      <th>CG17159</th>\n",
       "      <td>AACGAGAGTCCGGCTGAAAAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.691403</td>\n",
       "      <td>3.281687</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XHet@68786@-</th>\n",
       "      <th>FBgn0039945</th>\n",
       "      <th>CG17159</th>\n",
       "      <td>TAACGAGAGTCCGGCTGAAAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.476910</td>\n",
       "      <td>2.660218</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XHet@77281@+</th>\n",
       "      <th>FBgn0003559</th>\n",
       "      <th>su(f)</th>\n",
       "      <td>TCACACCTATCGTGGGCATCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.203569</td>\n",
       "      <td>0.420474</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YHet@328017@-</th>\n",
       "      <th>FBgn0085792</th>\n",
       "      <th>CR41509</th>\n",
       "      <td>GTTGAGGAAGCACTTGAAGAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.310891</td>\n",
       "      <td>3.658054</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dmel_mitochondrion_genome@14200@-</th>\n",
       "      <th>FBgn0013688</th>\n",
       "      <th>mt:srRNA</th>\n",
       "      <td>ACACATCGCCCGTCGCTCTTA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.693381</td>\n",
       "      <td>6.620448</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145144 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            ('INFO', 'motif_F10')  \\\n",
       "10@100177160@-                    ENSG00000107521 HPS1      CTAGACCACCCTCCATGTCAG   \n",
       "10@100177994@-                    ENSG00000107521 HPS1      TCCAGATGATCGAGGTGCCCG   \n",
       "10@100183394@-                    ENSG00000107521 HPS1      GGACTTCTTGCTGGTGAAGAG   \n",
       "10@100183566@-                    ENSG00000107521 HPS1      CAGCCCCCAGCAGGGGAGGCC   \n",
       "10@100184075@-                    ENSG00000107521 HPS1      CCGGATCCTCCTGGGAGTGAG   \n",
       "...                                                                           ...   \n",
       "XHet@68785@-                      FBgn0039945     CG17159   AACGAGAGTCCGGCTGAAAAG   \n",
       "XHet@68786@-                      FBgn0039945     CG17159   TAACGAGAGTCCGGCTGAAAA   \n",
       "XHet@77281@+                      FBgn0003559     su(f)     TCACACCTATCGTGGGCATCG   \n",
       "YHet@328017@-                     FBgn0085792     CR41509   GTTGAGGAAGCACTTGAAGAA   \n",
       "dmel_mitochondrion_genome@14200@- FBgn0013688     mt:srRNA  ACACATCGCCCGTCGCTCTTA   \n",
       "\n",
       "                                                            Human  Mouse  \\\n",
       "10@100177160@-                    ENSG00000107521 HPS1        1.0    NaN   \n",
       "10@100177994@-                    ENSG00000107521 HPS1        3.0    NaN   \n",
       "10@100183394@-                    ENSG00000107521 HPS1        3.0    NaN   \n",
       "10@100183566@-                    ENSG00000107521 HPS1        3.0    NaN   \n",
       "10@100184075@-                    ENSG00000107521 HPS1        3.0    NaN   \n",
       "...                                                           ...    ...   \n",
       "XHet@68785@-                      FBgn0039945     CG17159     NaN    NaN   \n",
       "XHet@68786@-                      FBgn0039945     CG17159     NaN    NaN   \n",
       "XHet@77281@+                      FBgn0003559     su(f)       NaN    NaN   \n",
       "YHet@328017@-                     FBgn0085792     CR41509     NaN    NaN   \n",
       "dmel_mitochondrion_genome@14200@- FBgn0013688     mt:srRNA    NaN    NaN   \n",
       "\n",
       "                                                            X. laevis  \\\n",
       "10@100177160@-                    ENSG00000107521 HPS1            NaN   \n",
       "10@100177994@-                    ENSG00000107521 HPS1            NaN   \n",
       "10@100183394@-                    ENSG00000107521 HPS1            NaN   \n",
       "10@100183566@-                    ENSG00000107521 HPS1            NaN   \n",
       "10@100184075@-                    ENSG00000107521 HPS1            NaN   \n",
       "...                                                               ...   \n",
       "XHet@68785@-                      FBgn0039945     CG17159         NaN   \n",
       "XHet@68786@-                      FBgn0039945     CG17159         NaN   \n",
       "XHet@77281@+                      FBgn0003559     su(f)           NaN   \n",
       "YHet@328017@-                     FBgn0085792     CR41509         NaN   \n",
       "dmel_mitochondrion_genome@14200@- FBgn0013688     mt:srRNA        NaN   \n",
       "\n",
       "                                                            X. tropicalis  \\\n",
       "10@100177160@-                    ENSG00000107521 HPS1                NaN   \n",
       "10@100177994@-                    ENSG00000107521 HPS1                NaN   \n",
       "10@100183394@-                    ENSG00000107521 HPS1                NaN   \n",
       "10@100183566@-                    ENSG00000107521 HPS1                NaN   \n",
       "10@100184075@-                    ENSG00000107521 HPS1                NaN   \n",
       "...                                                                   ...   \n",
       "XHet@68785@-                      FBgn0039945     CG17159             NaN   \n",
       "XHet@68786@-                      FBgn0039945     CG17159             NaN   \n",
       "XHet@77281@+                      FBgn0003559     su(f)               NaN   \n",
       "YHet@328017@-                     FBgn0085792     CR41509             NaN   \n",
       "dmel_mitochondrion_genome@14200@- FBgn0013688     mt:srRNA            NaN   \n",
       "\n",
       "                                                            Zebrafish  Fly  \\\n",
       "10@100177160@-                    ENSG00000107521 HPS1            NaN  NaN   \n",
       "10@100177994@-                    ENSG00000107521 HPS1            NaN  NaN   \n",
       "10@100183394@-                    ENSG00000107521 HPS1            NaN  NaN   \n",
       "10@100183566@-                    ENSG00000107521 HPS1            NaN  NaN   \n",
       "10@100184075@-                    ENSG00000107521 HPS1            NaN  NaN   \n",
       "...                                                               ...  ...   \n",
       "XHet@68785@-                      FBgn0039945     CG17159         NaN  2.0   \n",
       "XHet@68786@-                      FBgn0039945     CG17159         NaN  2.0   \n",
       "XHet@77281@+                      FBgn0003559     su(f)           NaN  2.0   \n",
       "YHet@328017@-                     FBgn0085792     CR41509         NaN  2.0   \n",
       "dmel_mitochondrion_genome@14200@- FBgn0013688     mt:srRNA        NaN  2.0   \n",
       "\n",
       "                                                                   X  \\\n",
       "10@100177160@-                    ENSG00000107521 HPS1      9.848524   \n",
       "10@100177994@-                    ENSG00000107521 HPS1      3.548068   \n",
       "10@100183394@-                    ENSG00000107521 HPS1      8.139991   \n",
       "10@100183566@-                    ENSG00000107521 HPS1      7.001106   \n",
       "10@100184075@-                    ENSG00000107521 HPS1      6.508685   \n",
       "...                                                              ...   \n",
       "XHet@68785@-                      FBgn0039945     CG17159   7.691403   \n",
       "XHet@68786@-                      FBgn0039945     CG17159   6.476910   \n",
       "XHet@77281@+                      FBgn0003559     su(f)     4.203569   \n",
       "YHet@328017@-                     FBgn0085792     CR41509   7.310891   \n",
       "dmel_mitochondrion_genome@14200@- FBgn0013688     mt:srRNA  9.693381   \n",
       "\n",
       "                                                                   Y  Cluster  \n",
       "10@100177160@-                    ENSG00000107521 HPS1      8.074862        1  \n",
       "10@100177994@-                    ENSG00000107521 HPS1      1.524192        3  \n",
       "10@100183394@-                    ENSG00000107521 HPS1      2.309283        3  \n",
       "10@100183566@-                    ENSG00000107521 HPS1      1.070967        3  \n",
       "10@100184075@-                    ENSG00000107521 HPS1      0.452521        3  \n",
       "...                                                              ...      ...  \n",
       "XHet@68785@-                      FBgn0039945     CG17159   3.281687        3  \n",
       "XHet@68786@-                      FBgn0039945     CG17159   2.660218        3  \n",
       "XHet@77281@+                      FBgn0003559     su(f)     0.420474        3  \n",
       "YHet@328017@-                     FBgn0085792     CR41509   3.658054        3  \n",
       "dmel_mitochondrion_genome@14200@- FBgn0013688     mt:srRNA  6.620448        1  \n",
       "\n",
       "[145144 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c351a6d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extact_all_fasta(df_in, column, fn_out, rna=True):\n",
    "    N = 0\n",
    "    with open(fn_out, \"w\") as output:\n",
    "        for idx, row in df_in.iterrows():\n",
    "            if rna == True:\n",
    "                output.write(\">{}\\n{}\\n\".format(N, row[column].replace(\"T\", \"U\")))\n",
    "            else:\n",
    "                output.write(\">{}\\n{}\\n\".format(N, row[column]))\n",
    "            N += 1        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8bed69e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "extact_all_fasta(df, motif_col, fasta_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0d7ccfc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_meme_file(df_in, id_column, motif_column, fn_out, rna=True):\n",
    "    all_ids = set(df_in[id_column].tolist())\n",
    "    temp_mat_name = fn_out+\".temp.mat\"\n",
    "    print(all_ids)\n",
    "    with open(temp_mat_name, \"w\") as output:\n",
    "        for ID in all_ids:\n",
    "            print(ID)\n",
    "            subdf = df_in[df_in[id_column]==ID]\n",
    "            count_data = {}\n",
    "            for _, row in subdf.iterrows():\n",
    "                if \"N\" in row[motif_column]:\n",
    "                    continue\n",
    "                if rna == True:\n",
    "                    iterseq = row[motif_column].replace(\"T\", \"U\")\n",
    "                else:\n",
    "                    iterseq = row[motif_column]\n",
    "                for idx, base in enumerate(list(iterseq)):\n",
    "                    if idx not in count_data:\n",
    "                        count_data[idx] = {\"A\":0, \"C\": 0, \"G\": 0, \"U\":0}\n",
    "                    count_data[idx][base] += 1\n",
    "            count_df = pd.DataFrame.from_dict(count_data).T\n",
    "            seqs = count_df.values\n",
    "            seqs = np.array(seqs)\n",
    "            logodata = LogoData.from_counts(counts=seqs, alphabet='ACGU')\n",
    "\n",
    "            temp = []\n",
    "            for i in range(logodata.counts.shape[0]):\n",
    "                # temp.extend(list(logodata.entropy[i]*logodata.counts[i]/logodata.counts[i].sum()))\n",
    "                output.write(\"{}\\t{}\\t{}\\t{}\\n\".format(logodata.counts[i][0], logodata.counts[i][1], logodata.counts[i][2], logodata.counts[i][3]))\n",
    "            output.write(\"\\n\")\n",
    "        \n",
    "    !matrix2meme -rna < $temp_mat_name > $fn_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d58fc365",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1, 2, 3}\n",
      "1\n",
      "2\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "generate_meme_file(df, cluster_col, motif_col, meme_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ab4e396f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Checking alphabets in 1 motif files.\n",
      "# Loading motifs from file 'mixed.meme'\n",
      "# Alphabet: RNA\n",
      "# NOTE: Will convert any DNA sequences to RNA.\n",
      "# Positive sequences \"mixed.fa\" - training: 130630 hold-out: 14514\n",
      "# Negative sequences are shuffled primary sequences (2-order) - training: 130630 hold-out: 14514\n",
      "# Estimating background model from control sequences.\n",
      "# Background: A 0.239 C 0.274 G 0.294 U 0.192\n",
      "# Background order: 2 Background size: 84\n",
      "# Using Fisher Exact test for p-values.\n",
      "# Computing q-values.\n",
      "#   Cannot estimate pi_0 accurately from fewer than 100 p-values.\n",
      "#   Total p-values = 3. Using pi_zero = 1.0.\n",
      "# Freeing storage...\n"
     ]
    }
   ],
   "source": [
    "!$sea -oc $sea_out -p $fasta_fn -m $meme_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10b3bc90",
   "metadata": {},
   "source": [
    "### It is better to read the html."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "24a1530d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sea = pd.read_csv(\"./{}/sea.tsv\".format(sea_out), header=0, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7dcffe41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RANK</th>\n",
       "      <th>DB</th>\n",
       "      <th>ID</th>\n",
       "      <th>ALT_ID</th>\n",
       "      <th>CONSENSUS</th>\n",
       "      <th>TP</th>\n",
       "      <th>TP%</th>\n",
       "      <th>FP</th>\n",
       "      <th>FP%</th>\n",
       "      <th>ENR_RATIO</th>\n",
       "      <th>SCORE_THR</th>\n",
       "      <th>PVALUE</th>\n",
       "      <th>LOG_PVALUE</th>\n",
       "      <th>EVALUE</th>\n",
       "      <th>LOG_EVALUE</th>\n",
       "      <th>QVALUE</th>\n",
       "      <th>LOG_QVALUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>mixed.meme</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NNNNNNNNNNCVRRDNNNNNN</td>\n",
       "      <td>NNNNNNNNNNCRGGRNNNNNN</td>\n",
       "      <td>99431.0</td>\n",
       "      <td>76.12</td>\n",
       "      <td>26674.0</td>\n",
       "      <td>20.42</td>\n",
       "      <td>3.73</td>\n",
       "      <td>0.00079</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-43008.41</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-43007.31</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-43007.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>mixed.meme</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NNNNNNNNNHCUCCANNNNNN</td>\n",
       "      <td>NNNNNNNNNHCUCCANNNNNN</td>\n",
       "      <td>15117.0</td>\n",
       "      <td>11.57</td>\n",
       "      <td>2300.0</td>\n",
       "      <td>1.76</td>\n",
       "      <td>6.57</td>\n",
       "      <td>4.70000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-5616.67</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-5615.57</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-5615.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>mixed.meme</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NNNNNKUDGCCAMWUGNNNNN</td>\n",
       "      <td>NNNNDKUWGCCAAAUGNNNNN</td>\n",
       "      <td>7871.0</td>\n",
       "      <td>6.03</td>\n",
       "      <td>2653.0</td>\n",
       "      <td>2.03</td>\n",
       "      <td>2.97</td>\n",
       "      <td>0.02200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1411.22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1410.12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1410.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td># SEA (Simple Enrichment Analysis): Version 5....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td># The format of this file is described at http...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td># sea -oc sea_out -p mixed.fa -m mixed.meme</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                RANK          DB   ID  \\\n",
       "0                                                  1  mixed.meme  3.0   \n",
       "1                                                  2  mixed.meme  1.0   \n",
       "2                                                  3  mixed.meme  2.0   \n",
       "3  # SEA (Simple Enrichment Analysis): Version 5....         NaN  NaN   \n",
       "4  # The format of this file is described at http...         NaN  NaN   \n",
       "5        # sea -oc sea_out -p mixed.fa -m mixed.meme         NaN  NaN   \n",
       "\n",
       "                  ALT_ID              CONSENSUS       TP    TP%       FP  \\\n",
       "0  NNNNNNNNNNCVRRDNNNNNN  NNNNNNNNNNCRGGRNNNNNN  99431.0  76.12  26674.0   \n",
       "1  NNNNNNNNNHCUCCANNNNNN  NNNNNNNNNHCUCCANNNNNN  15117.0  11.57   2300.0   \n",
       "2  NNNNNKUDGCCAMWUGNNNNN  NNNNDKUWGCCAAAUGNNNNN   7871.0   6.03   2653.0   \n",
       "3                    NaN                    NaN      NaN    NaN      NaN   \n",
       "4                    NaN                    NaN      NaN    NaN      NaN   \n",
       "5                    NaN                    NaN      NaN    NaN      NaN   \n",
       "\n",
       "     FP%  ENR_RATIO  SCORE_THR  PVALUE  LOG_PVALUE  EVALUE  LOG_EVALUE  \\\n",
       "0  20.42       3.73    0.00079     0.0   -43008.41     0.0   -43007.31   \n",
       "1   1.76       6.57    4.70000     0.0    -5616.67     0.0    -5615.57   \n",
       "2   2.03       2.97    0.02200     0.0    -1411.22     0.0    -1410.12   \n",
       "3    NaN        NaN        NaN     NaN         NaN     NaN         NaN   \n",
       "4    NaN        NaN        NaN     NaN         NaN     NaN         NaN   \n",
       "5    NaN        NaN        NaN     NaN         NaN     NaN         NaN   \n",
       "\n",
       "   QVALUE  LOG_QVALUE  \n",
       "0     0.0   -43007.31  \n",
       "1     0.0    -5615.57  \n",
       "2     0.0    -1410.12  \n",
       "3     NaN         NaN  \n",
       "4     NaN         NaN  \n",
       "5     NaN         NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04128d1c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
