<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>For advanced users &mdash; iMVP-utils 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Dimension Reduction Algorithms" href="UMAP.html" />
    <link rel="prev" title="For beginners" href="beginners.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> iMVP-utils
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="background.html">Background</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="beginners.html">For beginners</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">For advanced users</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#handling-stubborn-clusters">Handling stubborn clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#spike-in-imvp">Spike-in iMVP</a></li>
<li class="toctree-l2"><a class="reference internal" href="#phase-matching">Phase matching</a></li>
<li class="toctree-l2"><a class="reference internal" href="#approximate-clustering">Approximate clustering</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="UMAP.html">Dimension Reduction Algorithms</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="HDBSCAN.html">Unsupervised Clustering algorithms</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="GPU_accerleration.html">Towards huge datasets</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="limitations.html">Limitations</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks.html">Notebook indexes</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gallery.html">Gallery</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">iMVP-utils</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>For advanced users</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/advanced_users.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="for-advanced-users">
<h1>For advanced users<a class="headerlink" href="#for-advanced-users" title="Permalink to this headline"></a></h1>
<p>For advanced users, we suggest you to start with <code class="docutils literal notranslate"><span class="pre">HeLa_Noc.ipynb</span></code>, <code class="docutils literal notranslate"><span class="pre">Human_dev.ipynb</span></code>, and <code class="docutils literal notranslate"><span class="pre">m6A_methods.ipynb</span></code>. The following sections describe solutions for some special scenarios.</p>
<section id="handling-stubborn-clusters">
<h2>Handling stubborn clusters<a class="headerlink" href="#handling-stubborn-clusters" title="Permalink to this headline"></a></h2>
<p>If you find that the sequences in your data is hard to be separated. They are some options you can try:</p>
<ul class="simple">
<li><p>For the input, try to combine other features, like secondary structures with the sequences. Or try to use sequences longer or shorter than 21-mers.</p></li>
<li><p>For <code class="docutils literal notranslate"><span class="pre">UMAP</span></code>, increase <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> should help to some extents.</p></li>
<li><p>For <code class="docutils literal notranslate"><span class="pre">HDBSCAN</span></code>, if you have been adjusting the <code class="docutils literal notranslate"><span class="pre">min_samples</span></code> and <code class="docutils literal notranslate"><span class="pre">min_cluster_size</span></code> paramters for a long time, you can also try <code class="docutils literal notranslate"><span class="pre">cluster_selection_epsilon</span></code> parameter. However, we suggest you to observe your <code class="docutils literal notranslate"><span class="pre">condensed</span> <span class="pre">tree</span></code> and determine the if it is valuable to continue.</p></li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The k-mers of your sequences should always be a ultra small fraction in the k-mers space.</p>
</div>
<p>Here is a breif example for Noc-treated HeLa cells:</p>
<img alt="_images/HeLa_Noc.eom.png" class="align-center" src="_images/HeLa_Noc.eom.png" />
<p>This one is the initial UMAP clusters we produced with <code class="docutils literal notranslate"><span class="pre">eom</span></code> function. We can clearly find that the clusters in red and blue are well clustered but the green one has two centroids. These two sub-clusters are two adjacent to each other and therefore considered to be one in <code class="docutils literal notranslate"><span class="pre">HDBSCAN</span></code>. It is exhausted to adjust the paramters to separate them without generating too many small clusters. So here we can be smart to use the <code class="docutils literal notranslate"><span class="pre">leaf</span></code> method for sub-clustering.</p>
<p>Now we have two choices, directly plot the leaves of <strong>all</strong> sites, or just plot the leaves for <strong>cluster #3</strong>. Here we use the latter strategy.</p>
<img alt="_images/HeLa_Noc.eom_leaf.png" class="align-center" src="_images/HeLa_Noc.eom_leaf.png" />
<p>In this figure, we just keep cluster #1 and cluster #2 unchanged, but continue to split cluster #3. This generates the cluster #4 to #15, where #4 is what we want and #5 to #15 can be grouped into the orignally cluster #3. Then we have these beautiful clusters:</p>
<a class="reference internal image-reference" href="_images/HeLa_Noc.curated.png"><img alt="_images/HeLa_Noc.curated.png" class="align-center" src="_images/HeLa_Noc.curated.png" style="width: 350px;" /></a>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Saddly, HDBSCAN cannot use the lambda value to split the tree now.</p>
</div>
</section>
<section id="spike-in-imvp">
<h2>Spike-in iMVP<a class="headerlink" href="#spike-in-imvp" title="Permalink to this headline"></a></h2>
<p>Using “Spike-ins” (sequences with known modifications) is a quick way to find suitable clusters from the projections. Please read our <code class="docutils literal notranslate"><span class="pre">RNA_seq_var.ipynb</span></code> notebook for more information.</p>
</section>
<section id="phase-matching">
<h2>Phase matching<a class="headerlink" href="#phase-matching" title="Permalink to this headline"></a></h2>
<p>Owe to the position-sensitive feature of <code class="docutils literal notranslate"><span class="pre">iMVP</span></code>, we can have some interesting application.</p>
<p>In variant call from <code class="docutils literal notranslate"><span class="pre">Nanopre</span></code> RNA sequencing, an obvious problem is that the signal might not directly come from called base itself, but might come from its adjacent bases. Here, we call such phenomenon “phase mismatch”. In past, we might try to assign the signal of the variant to a selected adjacent base with some conditions. Now, with <code class="docutils literal notranslate"><span class="pre">iMVP</span></code>, we can have a more rational assignment because we now have the prior knowledge about the clusters. In our pratices, we firstly cluster the signals from <code class="docutils literal notranslate"><span class="pre">xPore</span></code>, and find out that the first four clusters were from m6A signals in different phasese, while an additional one is in a m6Am-like motif. Here, we fix the phase of cluster #1 to #4 that assign the sequences without an “A” base centered to a specific downstream or upstream bases, by the pattern of the clusters. With such a method, we can succeed to correct the phases of the m6A and m6Am-like motifs.</p>
<img alt="_images/phase_matching.png" class="align-center" src="_images/phase_matching.png" />
</section>
<section id="approximate-clustering">
<h2>Approximate clustering<a class="headerlink" href="#approximate-clustering" title="Permalink to this headline"></a></h2>
<p>When handling huge amount of sites, it is not easy for us to find a set of parameters perfectly clustering them. We, human beings, are also not good at finding clusters within millions of points. For most of the situations, like the global A-to-I RNA editing analysis, it is not necessary to assign very single point to a cluster. We want to have some snapshots. To solve this problem, we mimicks the density profiling step in <code class="docutils literal notranslate"><span class="pre">HDBSCAN</span></code> to generate a histogram of site density first. With the density histogram, we can isolate the clusters with computer visulization methods (e.g., functions in <code class="docutils literal notranslate"><span class="pre">openCV</span></code>). In our practices, we firstly perform noise reduction to the histogram to filter out the low density connections between the clusters. Then we use <code class="docutils literal notranslate"><span class="pre">find</span> <span class="pre">contours</span></code> functions to link the connected pixels. We finally extract the clusters from the histogram and annotate them to the orignal sites. Read notebooks in <code class="docutils literal notranslate"><span class="pre">RAPIDS</span></code> folder for more information.</p>
<img alt="_images/approximate_clustering.png" class="align-center" src="_images/approximate_clustering.png" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="beginners.html" class="btn btn-neutral float-left" title="For beginners" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="UMAP.html" class="btn btn-neutral float-right" title="Dimension Reduction Algorithms" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Jianheng Liu, Jing Yao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>