<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Dimension Reduction Algorithms &mdash; iMVP-utils 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Unsupervised Clustering algorithms" href="HDBSCAN.html" />
    <link rel="prev" title="For advanced users" href="advanced_users.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> iMVP-utils
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="background.html">Background</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="beginners.html">For beginners</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="advanced_users.html">For advanced users</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Dimension Reduction Algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#benchmarking">Benchmarking</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scanpy-and-imvp">Scanpy and iMVP</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parameters-of-umap">Parameters of UMAP</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="HDBSCAN.html">Unsupervised Clustering algorithms</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="GPU_accerleration.html">Towards huge datasets</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="limitations.html">Limitations</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks.html">Notebook indexes</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gallery.html">Gallery</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">iMVP-utils</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Dimension Reduction Algorithms</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/UMAP.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="dimension-reduction-algorithms">
<h1>Dimension Reduction Algorithms<a class="headerlink" href="#dimension-reduction-algorithms" title="Permalink to this headline"></a></h1>
<section id="benchmarking">
<h2>Benchmarking<a class="headerlink" href="#benchmarking" title="Permalink to this headline"></a></h2>
<img alt="_images/dimension_reduction.png" class="align-center" src="_images/dimension_reduction.png" />
<p>There are many kinds of algorithms for dimension reduction. In <strong>(a)</strong>, we use the m5C sites in fly embryos, which are validated to be majorly contributed by NSUN2 (blue) and NSUN6 (red), to test the peformace of four types of algorithms (PCA, LLE, t-SNE, and UMAP). As previously pointed out, UMAP is the best algorithm for condensing the sequences. t-SNE will break the relationships of the data points, and too sparse in some regions. PCA is somehow lucky in this test, but perform very bad with more complicated data (see below). DensMAP, a variant of UMAP, works well for some cases, but still not so suitable for compliated datasets (see below, too).</p>
<p>For <strong>(b)</strong> memory and time usages, we found that UMAP is moderate in computation burdern. Notably, UMAP requires a JIT (Just-in-Time) compiling process when it runs at the first time. This mechanism makes UMAP showhow slower in some cases. However, when JIT is finished, UMAP runs much faster than t-SNE.</p>
<p>In Noc-treated HeLa cells, which contains 4 types of motifs, PCA and DensMAP performed not so good. From the density contour drawn in the figure, we can clear found a cluster of Nop2 (purple) tightly adjacent to the cluster of NSUN2 (green). However, such a pattern is not clear in both PCA and DensMAP.</p>
<img alt="_images/UMAP_PCA_DensMAP.png" class="align-center" src="_images/UMAP_PCA_DensMAP.png" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="scanpy-and-imvp">
<h2>Scanpy and iMVP<a class="headerlink" href="#scanpy-and-imvp" title="Permalink to this headline"></a></h2>
<p>Here is a small disscussion about <code class="docutils literal notranslate"><span class="pre">Scanpy</span></code> (<a class="reference external" href="https://scanpy.readthedocs.io/en/stable/tutorials.html">https://scanpy.readthedocs.io/en/stable/tutorials.html</a>), a popular package for single cell analysis. In fact, we also tested Scanpy at the beginning, but we found that some assumptions in single cell analysis is different from that in modification motif analysis.</p>
<ol class="arabic simple">
<li><p>Normally we will perfrom a pre-processing on the single-cell data with PCA, to reduce the number of dimensions (yes, we can skip it in <code class="docutils literal notranslate"><span class="pre">Scanpy</span></code>). However, such a process cause the loss of information for UMAP. Normally, a series of one-hot encoded sequences should be perfect in practices, and such a pre-processing is not necessary.</p></li>
<li><p>The data structure packed in <code class="docutils literal notranslate"><span class="pre">Scanpy</span></code> is not so friendly in our analysis. We would love to have a more flexible one in pratices.</p></li>
</ol>
</section>
<section id="parameters-of-umap">
<h2>Parameters of UMAP<a class="headerlink" href="#parameters-of-umap" title="Permalink to this headline"></a></h2>
<p>Here are some tips for selecting <code class="docutils literal notranslate"><span class="pre">UMAP</span></code> parameters:</p>
<p><code class="docutils literal notranslate"><span class="pre">init</span></code> will impact the performance of <code class="docutils literal notranslate"><span class="pre">UMAP</span></code>, but that might not affect the clusters.</p>
<p><code class="docutils literal notranslate"><span class="pre">min_dist</span></code> is important to make sure that clusters will be separated correctly. We will use a low (close to zero) min_dist in practices.</p>
<p><code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> also affect the final result of projection. Higher <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> will make it runs longer. We suggest a value between 20 to 50. A super high <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> will not help your projection a lot.</p>
<p><code class="docutils literal notranslate"><span class="pre">metrics</span></code> is an important parameter for the reulsts. In practices, most of the metrics provied by <code class="docutils literal notranslate"><span class="pre">UMAP</span></code> have similar performance handling one-hot encoded sequences, except for <code class="docutils literal notranslate"><span class="pre">chbyshev</span></code>. Among these methods, the standard <code class="docutils literal notranslate"><span class="pre">eucilidean</span></code> (in CPU version) and <code class="docutils literal notranslate"><span class="pre">minkowski</span></code> (in cuML) work fast, so we suggest you to use the default one.</p>
<img alt="_images/UMAP_metrics.png" class="align-center" src="_images/UMAP_metrics.png" />
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Remember that the reproducibility of <code class="docutils literal notranslate"><span class="pre">UMAP</span></code> dependencies on both <code class="docutils literal notranslate"><span class="pre">random_state</span></code> and <code class="docutils literal notranslate"><span class="pre">n_jobs</span></code>, in addition to other key parameters.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="advanced_users.html" class="btn btn-neutral float-left" title="For advanced users" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="HDBSCAN.html" class="btn btn-neutral float-right" title="Unsupervised Clustering algorithms" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Jianheng Liu, Jing Yao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>